src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp

```diff
@@ -20,86 +25,125 @@ class BankManager {
+        auto allocated_tensor = ov::make_tensor(remote_tensor);
+        transformed_tensor.copy_to(allocated_tensor);
+        transformed_tensor = allocated_tensor;
+    }
+
+    return transformed_tensor;
 }
 
-ov::Tensor Bank::get(const ov::Tensor& tensor, const std::string& device) {
-    if (!tensor) {
-        OPENVINO_THROW("Uninitialized tensor in weights bank allocation!");
-    }
+void Bank::registerLT(const LazyTensor& tensor, const std::string& device) {
+    const std::string& device_for_alloc = m_alloc_device.empty() ? device : m_alloc_device;
 
-    if (device != "CPU" && device != "NPU") {
-        OPENVINO_THROW("Unsupported device in weights bank allocation: ", device);
+    if (device_for_alloc != "CPU" && device_for_alloc != "NPU") {
+        OPENVINO_THROW("Unsupported device in weights bank allocation: ", device_for_alloc);
     }
 
     std::lock_guard<std::mutex> guard(m_mutex);
 
-    auto iter_cpu = m_bank.find(tensor.data());
-    if (iter_cpu == m_bank.end()) {
-        OPENVINO_THROW("Unknown tensor in weights bank allocation!");
+    auto& device_bank = m_device_bank[device_for_alloc];
+    if (device_bank.find(tensor) == device_bank.end()) {
+        device_bank[tensor] = ov::Tensor();
     }
+}
 
-    // If target device is CPU - just reuse the default bank
-    if (device == "CPU") {
-        return iter_cpu->second;
-    }
+void Bank::evaluate_and_allocate() {
+    std::lock_guard<std::mutex> guard(m_mutex);
 
-    // Non-CPU - check if the tensor is already there
-    auto& device_bank = m_device_bank[device];
-    auto iter_device = device_bank.find(tensor.data());
-    if (iter_device != device_bank.end()) {
-        // Already allocated on the device - reuse
-        return iter_device->second;
+    for (auto&& bank : m_device_bank) {
+        const auto& device_for_alloc = bank.first;
+        auto& device_bank = bank.second;
+        std::vector<LazyTensor> vec;
+        for (const auto& el : device_bank) {
+            vec.push_back(el.first);
+        }
+        ov::parallel_for(vec.size(), [&](std::size_t idx) {
+            const auto& lt = vec[idx];
+            auto& transformed_tensor = device_bank[lt];
+            if (!transformed_tensor) {
+                transformed_tensor = lt.eval();
+            }
+
+            // Device name is known for each LT - set by registerLT
+            if (device_for_alloc != "CPU") {
+                // Allocate
+                m_remote_ctx = m_core->get_default_context(device_for_alloc)._ptr;
+                auto remote_tensor = m_remote_ctx->create_host_tensor(transformed_tensor.get_element_type(),
+                                                                      transformed_tensor.get_shape());
+                auto allocated_tensor = ov::make_tensor(remote_tensor);
+                transformed_tensor.copy_to(allocated_tensor);
+                transformed_tensor = allocated_tensor;
+            }
```

Comment:
It seems this code is called even if it was already done before (tensor in the device bank is already in the right memory)

Reply:
Done